confluent_kafka/consumer.py

```diff
@@ -0,0 +1,145 @@
+        calling ```client.poll()``` or ```client.flush()```.
+
+        :param dict conf: Configuration properties.
+            See https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md for more information.
+        :param func key_deserializer(topic, key): Converts message key bytes to object.
+            **note** deserializers are responsible for handling NULL keys
+        :param func value_deserializer(topic, value): Converts message value bytes to object.
+            **note** deserializers are responsible for handling NULL values
+        :param func on_commit(err, [partitions]): Callback used to indicate success or failure
+            of an offset commit.
+        :param func stats_cb(json_str): Callback for statistics emitted every ``statistics.interval.ms``.
+            See https://github.com/edenhill/librdkafka/wiki/Statistics” for more information.
+        :param func throttle_cb(confluent_kafka.ThrottleEvent): Callback for throttled request reporting.
+        :param logging.handlers logger: Forwards logs from the Kafka client to the provided handler instance.
+            Log messages will only be forwarded when ``client.poll()`` or ``producer.flush()`` are called.
+        :raises TypeError: If conf is not a dict.
+    """
+
+    __slots__ = ["_key_deserializer", "_value_deserializer"]
+
+    # conf must remain optional as long as kwargs are supported
+    def __init__(self, conf={}, key_deserializer=None, value_deserializer=None,
+                 on_commit=None, stats_cb=None, throttle_cb=None, logger=None, **kwargs):
+
+        if not isinstance(conf, dict):
+            raise TypeError("expected configuration dict")
+
+        if kwargs:
+            # Handle kwargs for backwards compatibility
+            conf.update(kwargs)
+            warn("The use of kwargs is being deprecated. "
+                 "In future releases `conf` will be mandatory and "
+                 "all keyword arguments must match the constructor signature explicitly.",
+                 category=DeprecationWarning, stacklevel=2)
+
+        self._key_deserializer = key_deserializer
+        self._value_deserializer = value_deserializer
+
+        # Callbacks can be set in the conf dict or *ideally* as parameters.
+        # Handle both cases prior to passing along to _impl
+        # If callbacks are configured in both places parameter values take precedence.
+        if not on_commit:
+            on_commit = conf.get('on_commit', None)
+
+        if not stats_cb:
+            stats_cb = conf.get('stats_cb', None)
+
+        if not throttle_cb:
+            throttle_cb = conf.get('throttle_cb', None)
+
+        if not logger:
+            logger = conf.get('logger', None)
+        super(Consumer, self).__init__(conf, on_commit=on_commit, stats_cb=stats_cb,
+                                       throttle_cb=throttle_cb, logger=logger)
+
+    def poll(self, timeout=-1.0, key_deserializer=None, value_deserializer=None):
```

Comment:
We can't maintain compatibility with the existing AvroProducer/Consumer without it. Since we have already provided this in the API we can't just rip it out without notice.

Reply:
this is a new class - i don't think ensuring users can swap out the old avro producer and consumer for this one has any priority - you're already breaking compatibility by requiring they change classes.

Reply:
The idea actually is to get away from having a specific class for Avro. Instead this should be configurable using the serializer interface proposed in this PR. You are right though per request overrides don't make sense in the Consumer. You can't actually pass your own schema per `poll()` in the AvroConsumer either. I just got ahead of myself.

Reply:
Why are AvroProducer/AvroConsumer a bad idea?

Reply:
>this is a new class 

Thank you for this @mhowlett , you are right and this got me thinking. The avro producer isn't actually allowing for serializer overrides so much as it's allowing you to optionally introduce a new datum writer by providing a new schema. Either way this is an example of the serializer's responsibility leaking into the producer itself which is not what we want. Instead the serializer should handle this. What makes this weird for Avro is that the python implementation does not actually have this notion of `Specific` or even `Generic` record types. If you want a record you pass a dict with no schema tied to it. That said this is a new class, and a new `GenericRecord` serializer(and supporting GenericRecord type (binding schema to dict)  can also be introduced 
 to handle this. In the end this will make for a much better Avro experience for everyone while also removing all these checks from the produce path.

Reply:
re: @edenhill's question. At one point with the .NET client, I was seriously considering separate producer/consumer and avro producer/consumer classes. The motivation for this was the observation that the serde interfaces are very simple, except for the requirements of the SR integration. It seemed reasonable to have P/Cs for simple serdes (which furthermore allowed for beautiful terse inline specification of them), which was applicable to most scenarios and defer avro functionality to elsewhere. Initially, i was thinking this could even be done with C# extension methods, not requiring a the classes for avro (but that was not the case). But it turned out to work much better to have completely general serde delegate types (two of them, one async, one not) and have the structure @rnpridgeon is suggesting above. @rnpridgeon's comment about avro being too specific also re-enforces this choice. 

@rnpridgeon - don't quite follow what you're saying, but broadly sounds good and look forward to reviewing it.